src/BungeeApproveAndBridge.sol

```diff
@@ -0,0 +1,167 @@
+/// 2. Modify output amount in calldata
+contract BungeeApproveAndBridge is ApproveAndBridge {
+    error InvalidInput();
+    error PositionOutOfBounds();
+    error BridgeFailed();
+
+    struct ModifyCalldataParams {
+        uint256 inputAmountIdx;
+        bool modifyOutput;
+        uint256 outputAmountIdx;
+        uint256 additionalValue;
+    }
+
+    uint8 private constant EXTRA_DATA_PARAMS_COUNT = 4;
+    uint8 private constant EXTRA_DATA_LENGTH_BYTES = 32;
+    uint8 private constant EXTRA_DATA_LENGTH = EXTRA_DATA_PARAMS_COUNT * EXTRA_DATA_LENGTH_BYTES;
+
+    ISocketGateway immutable socketGateway;
+
+    constructor(ISocketGateway socketGateway_) {
+        socketGateway = socketGateway_;
+    }
+
+    function bridgeApprovalTarget() public view override returns (address) {
+        return address(socketGateway);
+    }
+
+    function bridge(IERC20 token, uint256 amount, bytes calldata data) internal override {
+        // decode & parse data to find positions in calldata to modify
+        (bytes memory modifiedCalldata, uint256 additionalValue) = _parseAndModifyCalldata(amount, data);
+
+        // execute using the modified calldata via SocketGateway.fallback()
+        (bool success,) = address(token) == NATIVE_TOKEN_ADDRESS
+            ? address(socketGateway).call{value: amount + additionalValue}(modifiedCalldata)
+            : address(socketGateway).call{value: additionalValue}(modifiedCalldata);
+        if (!success) revert BridgeFailed();
+    }
+
+    function _parseAndModifyCalldata(uint256 amount, bytes calldata data)
+        internal
+        pure
+        returns (bytes memory, uint256)
+    {
+        // Parse the data into route calldata and extra data
+        (bytes memory routeCalldata, ModifyCalldataParams memory modifyCalldataParams) = _parseCalldata(data);
+
+        // Read the original input amount from the calldata
+        uint256 originalInput = _readUint256({_data: routeCalldata, _index: modifyCalldataParams.inputAmountIdx});
+
+        // Replace the input amount in the calldata
+        bytes memory modifiedCalldata =
+            _replaceUint256({_original: routeCalldata, _start: modifyCalldataParams.inputAmountIdx, _amount: amount});
+
+        // Optionally replace the output amount if required
+        // in case of bridges like Across, need to modify both input and output amounts
+        // - decode current input and output amounts from calldata
+        // - calculate and apply the percentage diff bw new and old input amount on the old output amount
+        // - replace the output amount at the index with the new amount
+        // - assumes output amount is always uint256 in SocketGateway impls
+        if (modifyCalldataParams.modifyOutput) {
+            uint256 originalOutput = _readUint256({_data: routeCalldata, _index: modifyCalldataParams.outputAmountIdx});
+            uint256 newOutput = _applyPctDiff({_base: originalInput, _compare: amount, _target: originalOutput});
```

Comment:
Will across and others using this respect the same price? Any chance they won't bridge because the amounts change?

Reply:
Haven't seen Across refusing to bridge with this. But will have to check and confirm for each bridge tho. eg. bridges that have an off-chain quote